### DISCUSSION TITLE
[Chapter 10] Redis

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/spring-boot-for-msa#156] [Chapter 10. Redis and Spring Cache] [answered]
Q1. 
레디스를 key-value 데이터 구조에 기반한 데이터베이스라고 처음에 소개하는데, 

RDBMS랑 key-value 외에도 많이 활용하고 있는 데이터베이스는 어떤 것들이 있을까? 그리고 실제로 활용하고 있는 사례가 있는가?

Q2. 
레디스를 이론과 실습으로만 익혀서 아직은 잘 감이 안오는데.. 실제로 적용해본 분들에게 궁금하다.!
책에서는 레디스를 사용할 수 있는 여러 라이브러리 중 Lettuce 에 대해 중점적으로 설명한다.

Lettuce 를 실제로 많이 사용하는가?   여러 라이브러리의 차이점이 있을까?
그리고 레디스의 장점으로 분산락을 설명하는데, 다른 많은 데이터베이스에서는 동시 접근을 어떻게 해결하는지 ?

Q3.
Redis 아키텍처에 대해 조금이나마 이해할 수 있었던 시간이었다.
이 책을 읽으면서 느낀 건 Redis를 그냥 비관계형 데이터베이스로만 생각했었는데,  
Kafka와 유사한 점이 많다고 생각이 들었다. (아키텍처 구조나 Pub/Sub 등)

**Comment by corock** [comment, marked as answer]
Q1. 스프링 공식 사이트에서 소개하는 [Spring Data MongoDB](https://spring.io/projects/spring-data-mongodb) 가 가장 먼저 생각나네요.
MongoDB 는 대표적인 NoSQL Document DB 중 하나로, 스키마가 유연하고 세션이나 로그 데이터 적재 시 유용하다고 알려져 있습니다.

실무에서는 캐싱 목적으로 MongoDB 를 활용하는 건 본 적이 있는데... 좋은 방법인지는 잘 모르겠습니다.
그럼에도 도입하고 있는 기업이 있으니, 아래 링크를 참고하면 좋을 것 같습니다.

1. [라인은 왜 몽고DB를 도입했을까](https://www.bloter.net/news/articleView.html?idxno=30335)
2. [무신사 서비스에 적합한 NoSQL 도입 여정 — 1편](https://medium.com/musinsa-tech/database-6d1052ca6b36)
3. [[드림어스컴퍼니 블로그] MongoDB - 도입기](https://www.blog-dreamus.com/post/flo-tech-mongodb-%EB%8F%84%EC%9E%85%EA%B8%B0)


Q2. 실무에서도 레디스 클라이언트로 Jedis 보다 Lettuce 를 많이 사용하는 것으로 보입니다.
Jedis 랑 비교해 보자면... Lettuce 는 다음과 같은 장점이 있다고 합니다.

1. Slave(Replication) DB 를 통한 조회 시 구현체 제공
2. 레디스 클러스터를 사용한 비동기 호출 지원
3. 트리거 기반 자동 [토폴로지 새로 고침](https://lettuce.io/core/release/api/io/lettuce/core/cluster/ClusterTopologyRefreshOptions.Builder.html) 지원

(분산 락은 10.5 에서 봅시다~)

Q3. 레디스는 데이터베이스고 카프카는 메시지 브로커라 용도가 다르나, 사용하는 기술에서 겹치는 부분이 있을 것 같긴 하네요.

**Reply by JoisFe** [reply]
레디스를 메시지 브로커로 쓰는 경우도 많더라구요

- [참고자료](https://aws.amazon.com/ko/compare/the-difference-between-rabbitmq-and-redis/)
- [참고자료 REDIS](https://redis.com/solutions/use-cases/messaging/)

**Comment by JoisFe** [comment]
- Spring Data Neo4J -> 그래프 디비
- [참고자료](https://spring.io/projects/spring-data-neo4j)
